### PR TITLE
Feature: misc functions on literals

### LINKED ISSUE #97: Feature: 17.4.4 Functions on Numerics
https://www.w3.org/TR/sparql11-query/#func-numerics

- [x] 17.4.4.1 abs
- [x] 17.4.4.2 round
- [x] 17.4.4.3 ceil
- [x] 17.4.4.4 floor
- [x] 17.4.4.5 RAND

### LINKED ISSUE #134: Enhancement: provide an efficient way to compare the datatype of literals with another datatype
Currently to compare the datatype of literals one has to

```c++
Literal lit = ...unknown literal...;
bool dt_eq = lit.datatype() == IRI{datatypes::xsd::String::identifier};
```
Which potentially involves 2 unecessary backend calls. Instead a function should provided that makes use of DatatypeIds to compare.

```c++
struct Literal {
    template<LiteralDatatype T>
    bool is() const noexcept;

    bool is(IRI const &datatype) const noexcept;
};
```

### LINKED ISSUE #136: Perf Bug: Literal::compare_impl should check backend handles for early exit
if the handles are equal we can do an early exit with `equal` which is currently not done

### PR DESCRIPTION
**Pull request by liss-h**
This PR resolves #97 and #99 (at least partially).

Currently missing:
- [`ENCODE_FOR_URI`](https://www.w3.org/TR/sparql11-query/#func-encode)
- ~~[`langMatches`](https://www.w3.org/TR/sparql11-query/#func-langMatches)~~

Other notes:
- ~~the regex matching is currently implemented with `std::regex`~~

Also added some user defined literal/iri/bnode operators to make literal construction less verbose

### CODE REVIEW

File: src/rdf4cpp/rdf/Literal.hpp
Comment (liss-h) [inline src/rdf4cpp/rdf/Literal.hpp:752]: What's your opinion on the two `substr` overloads having different behaviour?
- `substr(size_t, size_t, NodeStorage &)` behaves like `std::string_view::substr`
- `substr(Literal const &, Literal const &, NodeStorage &)` behaves as per [XPATH spec](https://www.w3.org/TR/xpath-functions/#func-substring) (you should read it, its weird 😅)

Note: my intention for having two overloads for each function was to have one version that is useful when working in "C++-land" and one version that is useful in "RDF-land" e.g. for query evaluation.

Argument for keeping is this way:
The XPATH spec is weird and the first overload is supposed to be useful in "C++-land" and a `substr(double, double)` is not what anyone would expect from a `substr` function in the context of C++. Especially not the 1-based indexing and handling of negative values.

Argument for making them consistent:
Well they are consistent then, but then we have to choose one behaviour and either move the XPATH weirdness into query evaluation or live with not having a proper C++y interface here.

Comment (bigerl) [reply src/rdf4cpp/rdf/Literal.hpp:752]: Let's keep both substr versions. I agree the definition is wild. Overhead seems little.

**Comment by bigerl** [comment]
Question: Does substr in xsd count unicode characters or `char`? Unicode characters can span multiple `char`s.

**Comment by liss-h** [comment]
> Question: Does substr in xsd count unicode characters or `char`? Unicode characters can span multiple `char`s.

"Definition: A character is an atomic unit of text as specified by ISO/IEC 10646:2000 [[ISO/IEC 10646]](https://www.w3.org/TR/REC-xml/#ISO10646). Legal characters are tab, carriage return, line feed, and the legal characters of Unicode and ISO/IEC 10646. [...]"

"any Unicode character, excluding the surrogate blocks, FFFE, and FFFF."

So I guess it counts `Unicode Scalar Values` (?)

Not sure but maybe this function works: [mblen](https://en.cppreference.com/w/c/string/multibyte/mblen)

**Comment by liss-h** [comment]
~~Currently waiting on #122 merge to be able to implement `langMatches`.~~ 👍

**Comment by liss-h** [comment]
Using utfcpp for unicode stuff now.
Also considered icu but it had problems with building on clang and the interface is very different to stdlib and also ineffiecient for our usecase.

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp]: Is this used anywhere?

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp]: Is this used anywhere (except in get_supertype_conversion below)?

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp:281]: Is this used anywhere (except in get_supertype_conversion below)?

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/owl/Rational.cpp]: This looks strange. Given r = 2/4 (=.5); divisor would be gcd(2,4) = 2- so 2/4 simplified would be 2/4/2 (= .25). 

Or am I overlooking sth?

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/owl/Rational.cpp:90]: I don't get why the rest thing is necessary. Shouldn't casting to integral simply result in an ceil already?

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/owl/Rational.cpp]:
```c++
auto const integral_part = static_cast<boost::multiprecision::cpp_int>(operand);
```

as below?

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/datatypes/owl/Rational.cpp:71]: Why do rational round and ceil have custom implementations instead of using boost functions like in real?

**Review by bigerl — CHANGES_REQUESTED** [review]
Partial review: abs, round, floor, ceil 

Thx for the PR. Some  things are unclear to me. Please review the comments below.

File: src/rdf4cpp/rdf/regex/Regex.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/regex/Regex.hpp]: Compilation Flag sounds like this decides if the regex is (pre-)compiled/optimized. I would suggest something like "flags to configure the regex mode".

File: src/rdf4cpp/rdf/regex/private/RegexReplacerImpl.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/regex/private/RegexReplacerImpl.cpp]: unsigned underflow is well defined in C++ and string_view's npos builds on it. So I would suggest to use it here. 

```c++
while (pos != std::string_view::npos) {
        if (pos == 0UL) {
            // is at start, cannot be escaped => replace
            replace(res, 0UL);
        } else {
            // find start of potential backslash sequence preceding the '$'
            auto const end = pos;
            auto before_start = end - 1UL;
            while (before_start != std::string_view::npos && res[before_start] == '\\') {
                --before_start;
            }

            if (auto const count = end - (before_start + 1UL); count % 2UL == 0UL) {
                // even number of '\' preceding '$' => not escaped => replace
                replace(res, pos);
            } else {
                // uneven number of '\' preceding '$' => escaped => remove escape char as RE2 does not require to escape '$'
                res.erase(pos - 1UL, 1UL);
                pos -= 1UL;
            }
        }

        pos = res.find_first_of('$', pos + 1UL);
    }
```

File: src/rdf4cpp/rdf/util/private/Utf8.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/util/private/Utf8.hpp]: I guess it is time now to introduce an folder private at the root level that contains the (relevant parts of the) same folder structure. Things that are used only in cpp files should live there. This also applies to parts of the parser, I think.

File: src/rdf4cpp/rdf/BlankNode.cpp
Comment (bigerl) [inline src/rdf4cpp/rdf/BlankNode.cpp]: I like the idea of string literals but the name literals is here a bit ambiguous. How about shorthands?

File: src/rdf4cpp/rdf/Literal.hpp
Comment (bigerl) [inline src/rdf4cpp/rdf/Literal.hpp]: No

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/owl/Rational.cpp]: No, you are right. The idea was dividing both numerator and denominator by divisor. Don't ask me why I thought these were equivalent 😅

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/owl/Rational.cpp:71]: because boost does not provide specializations of these functions for `rational`

File: src/rdf4cpp/rdf/datatypes/owl/Rational.cpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/owl/Rational.cpp:90]: ~~Casting to integral is usually a floor not a ceil. Thats why the rest thingy is there~~ Oh ok I think you meant that and just wrote ceil by accident. Yes makes sense

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp:281]: I'm not sure what I originally used that for but it appears unused now.

File: tests/nodes/tests_Literal.cpp
Comment (bigerl) [inline tests/nodes/tests_Literal.cpp]: please add a short comment that this check is non-deterministic. Will most probably never fail but should be documented.

File: tests/nodes/tests_Literal.cpp
Comment (bigerl) [inline tests/nodes/tests_Literal.cpp]: same as above

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp:281]: But I guess its a maybe useful accessor if you only want the table. If every other capability gets it's own accessor I guess this makes sense to have (?)

File: src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp]: Not sure what the original usage for that was, but appears unused.

File: tests/nodes/tests_Literal.cpp
Comment (bigerl) [inline tests/nodes/tests_Literal.cpp:683]: What exactly are the problems here? If it turns out to be complex, I would suggest to create an issue, fix the other things in ths PR and merge.

File: tests/util/tests_Regex.cpp
Comment (bigerl) [inline tests/util/tests_Regex.cpp:73]: Also something for an issue?

**Review by bigerl — CHANGES_REQUESTED** [review]
Review of the rest. 
This covers quite a lot: Regex, functions on strings, C++ literals (very nice idea BTW) and additions to CowString. Great work. All looks pretty well already. Some points that still need a little attention are listed below.

File: src/rdf4cpp/rdf/regex/Regex.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/regex/Regex.hpp]: But the regex does get pre-compiled. See https://github.com/google/re2/wiki/CplusplusAPI#pre-compiled-regular-expressions

File: tests/util/tests_Regex.cpp
Comment (liss-h) [reply tests/util/tests_Regex.cpp:73]: This shouldn't work because it's an invalid escape (even in RE2 syntax as far as I can tell).
RE2 just accepts it and does some weird things here. I probably need to change the translation somehow to catch these cases manually.

Comment (liss-h) [reply tests/util/tests_Regex.cpp:73]: I.e. with regex r = `[0-9]` this passes:

```
std::string s = "Hello 1 World";
r.make_replacer(R"(Hello \ World)").regex_replace(s);
CHECK(s == "Hello Hello  World");
```

File: tests/nodes/tests_Literal.cpp
Comment (liss-h) [reply tests/nodes/tests_Literal.cpp:683]: This is basically behaviour coming from RE2 and I'm not sure how to fix it from the outside.

Comment (liss-h) [reply tests/nodes/tests_Literal.cpp:683]: opened an issue #132

File: tests/util/tests_Regex.cpp
Comment (bigerl) [reply tests/util/tests_Regex.cpp:73]: I would guess it is an escaped space. Like in Latex `\ `.

File: src/rdf4cpp/rdf/regex/Regex.hpp
Comment (bigerl) [reply src/rdf4cpp/rdf/regex/Regex.hpp]: Yes but to me the phrase "compilation flag" sounds as if this flag would control if the regex gets pre-compiled.

The relevant part is that this flag decides if the resulting rdf4cpp Regex object will work in the modes DotAll, CaseInsensitive and/or Literal. The following is a guess: The flags would also be applicable if the backend regex implementation would not be used in pre-compiled mode.

File: tests/util/tests_Regex.cpp
Comment (liss-h) [reply tests/util/tests_Regex.cpp:73]: Do you think the behaviour is fine this way or is it important that we catch that?

Comment (liss-h) [reply tests/util/tests_Regex.cpp:73]: I just noticed something. Everything after a `\` which is not part of a reference to the match in the replacement string just gets deleted. You can see it in the example I gave above and also:

```
std::string s = "0";
r.make_replacer(R"(Hello \hgfhf World)").regex_replace(s);
CHECK(s == "Hello ");
```

It doesn't seem to matter what comes after the backslash if it isnt another backslash or a number the rest of the replacement just gets ignored. Why

Comment (bigerl) [reply tests/util/tests_Regex.cpp:73]: That is very odd. I would say we have for now some preliminary regex support. We state that the syntax follows google RE2. So users are warned. I would first complete features and then come back to this kind of fine-tuning. So, please document it in a issue and then mark this conversation as resolved.

File: src/rdf4cpp/rdf/util/private/Utf8.hpp
Comment (liss-h) [reply src/rdf4cpp/rdf/util/private/Utf8.hpp]: Do you mean

```
- src
    - rdf4cpp
    - private
```

Or
```
- src
    - rdf4cpp
        - private
```

?

Comment (bigerl) [reply src/rdf4cpp/rdf/util/private/Utf8.hpp]: I mean 
./src/rdf4cpp/ ...
./private/rdf4cpp/ ...

**Comment by liss-h** [comment]
This PR now also addresses #136 #134

**Review by bigerl — APPROVED** [review]
Changes for https://github.com/rdf4cpp/rdf4cpp/issues/136 and https://github.com/rdf4cpp/rdf4cpp/issues/134 look fine

**Review by bigerl — CHANGES_REQUESTED** [review]
Ready to merge after conflicts with develop are resolved

File: tests/util/tests_Regex.cpp
Comment (liss-h) [reply tests/util/tests_Regex.cpp:73]: #141

**Review by bigerl — APPROVED** [review]

**Pull request merged**
